File: android/app/src/main/java/com/dibay/app/nativevideo/NativeVideoCallOwner.java
package com.dibay.app.nativevideo;

import java.util.concurrent.ConcurrentHashMap;

/** Native Video single owner. Web/V4 may not claim calls owned here. */
public final class NativeVideoCallOwner {
  private static final ConcurrentHashMap<String, String> OWNERS = new ConcurrentHashMap<>();
  private static final ConcurrentHashMap<String, Boolean> TERMINAL = new ConcurrentHashMap<>();

  private NativeVideoCallOwner() {}

  public static boolean claimNative(String callId, String reason) {
    if (callId == null || callId.trim().isEmpty()) return false;
    String sid = callId.trim();
    if (TERMINAL.containsKey(sid)) {
      NativeVideoCallLog.warn("duplicate_runtime_blocked", sid, "reason=terminal_call_replay");
      return false;
    }
    String prev = OWNERS.putIfAbsent(sid, "native_video");
    if (prev == null) {
      NativeVideoCallLog.info("owner_claimed_native_video", sid, "reason=" + safe(reason));
      return true;
    }
    NativeVideoCallLog.warn("duplicate_runtime_blocked", sid, "reason=already_owned_native_video");
    return false;
  }

  public static boolean isNativeOwned(String callId) {
    return callId != null && "native_video".equals(OWNERS.get(callId.trim()));
  }

  public static void release(String callId, String reason) {
    if (callId == null || callId.trim().isEmpty()) return;
    String sid = callId.trim();
    String prev = OWNERS.remove(sid);
    TERMINAL.put(sid, Boolean.TRUE);
    NativeVideoCallLog.info("owner_released", sid, "owner=" + prev + " reason=" + safe(reason));
  }

  private static String safe(String value) {
    return value != null && !value.trim().isEmpty() ? value.trim() : "unknown";
  }
}

File: ios/App/App/Call/Video/NativeVideoCallOwner.swift
import Foundation

/**
 * Phase B0 — Native Video single owner skeleton.
 * Ports Android `NativeVideoCallOwner`. Web/V4 may not claim calls owned here.
 */
enum NativeVideoCallOwner {
  private static let lock = NSLock()
  private static var owners: [String: String] = [:]
  private static var terminal: Set<String> = []

  static func claimNative(callId: String, reason: String) -> Bool {
    let sid = callId.trimmingCharacters(in: .whitespacesAndNewlines)
    guard !sid.isEmpty else { return false }

    lock.lock()
    defer { lock.unlock() }

    if terminal.contains(sid) {
      NativeVideoCallLog.warn("duplicate_runtime_blocked", callId: sid, details: "reason=terminal_call_replay")
      return false
    }
    if owners[sid] != nil {
      NativeVideoCallLog.warn("duplicate_runtime_blocked", callId: sid, details: "reason=already_owned_native_video")
      return false
    }
    owners[sid] = "native_video"
    NativeVideoCallLog.info("owner_claimed_native_video", callId: sid, details: "reason=\(safe(reason))")
    return true
  }

  static func isNativeOwned(callId: String) -> Bool {
    let sid = callId.trimmingCharacters(in: .whitespacesAndNewlines)
    guard !sid.isEmpty else { return false }
    lock.lock()
    defer { lock.unlock() }
    return owners[sid] == "native_video"
  }

  static func release(callId: String, reason: String) {
    let sid = callId.trimmingCharacters(in: .whitespacesAndNewlines)
    guard !sid.isEmpty else { return }

    lock.lock()
    defer { lock.unlock() }

    let prev = owners.removeValue(forKey: sid)
    terminal.insert(sid)
    NativeVideoCallLog.info(
      "owner_released",
      callId: sid,
      details: "owner=\(prev ?? "none") reason=\(safe(reason))"
    )
  }

  private static func safe(_ value: String) -> String {
    let trimmed = value.trimmingCharacters(in: .whitespacesAndNewlines)
    return trimmed.isEmpty ? "unknown" : trimmed
  }
}
